### DISCUSSION TITLE
Callbacks on events

### DISCUSSION THREAD

**Discussion by wuyongzheng** [esp8266/Arduino#8468] [Ideas]
This is a feature request for ESP8266HTTPUpdateServer.

Background:
I'm using esp8266 to control WS2812 light strips. I want to turn off the light strip during update. Before ESP.restart(), I want to set some pins to LOW and some pins mode to INPUT.

Proposal:
It would be nice to support callback functions on various events such as
1. begin OTA
2. after OTA and before restart.

**Comment by amitesh-singh** [comment]
Call `ESPhttpUpdate.rebootOnUpdate(false);`
And then after update, do your stuffs and ESP.restart() afterwards.

**Reply by wuyongzheng** [reply]
Thanks for the reply. I'm using ESP8266HTTPUpdateServer, not ESPhttpUpdate. In my scenario, ESP8266 is the web server. I use a web browser to upload the firmware to the ESP. As far as I understand, ESPhttpUpdate works the opposite way, where ESP, the the web client downloads the firmware.

I think ESP8266HTTPUpdateServer doesn't respect ESPhttpUpdate.rebootOnUpdate(false). From the code referenced below, It calls ESP.restart() immediately upon successful update.
https://github.com/esp8266/Arduino/blob/9f536e68f8f3aa3ffae26493dd0a31e1957d2e81/libraries/ESP8266HTTPUpdateServer/src/ESP8266HTTPUpdateServer-impl.h#L73
